### DISCUSSION TITLE
Common key combos

### DISCUSSION THREAD

**Discussion by giatmanos** [lostallmymoney/Razer_Mouse_Linux#35] [Q&A] [answered]
Hi,
First of all thanks to @lostallmymoney for the amazing work! I have been using this for a long time now and I am very happy with all the features. Especially with all the updates lately. 

However, due to one of the latest updates (I think) something must have changed and my previous keyMap.txt common configuration does not work as expected anymore. I have set up macros for common copy paste (Ctrl+C/V) for my defaultconfig and they no longer work. I use e.g.:
```
1 - key=Control_L
1 - key=c
```
Is this the right way to configure for this macro or not?

**Comment by lostallmymoney** [comment]
Edit :
Fix to copy :
2 - keyPressOnPress=Control_L
2 - sleep=50
2 - keyClick=c
2 - sleep=50
2 - keyReleaseOnPress=Control_L

It is working on my end.
Are you sure you added it to the right config ?
There is a config specifically for Gnome-Terminal now for example !

You can also verify if you are using Wayland (sometimes ubuntu restarts on wayland for no reasons) which would mean using DOTOOL : https://git.sr.ht/~geb/dotool
Which means maybe different key names.
I'd recommend X11 for now.
I will update dotool right now too in case it might help you.
If you still do not succeed you can try using a command that will do the copy and pasting for you run=copy run=paste.
And if you still have an issue you can come back to me. Thanks for the feedback

**Reply by lostallmymoney** [reply]
Edit : Seems to work for a while until it doesn't. I think it might be a xdotool bug. I was already thinking of swapping all possible keys into specialKeys equivalent to bypass this.
You might be able to use xclip in the meanwhile.
I will look into this in the next week. Maybe in a few days.

**Reply by giatmanos** [reply]
It does indeed work sometimes but not regularly. I have found a better way which works with almost 100% consistency and is like so:
1 - key=Control_L
1 - keyClickRelease=c

**Reply by lostallmymoney** [reply]
The real fix might only be to create new versions of specialPress and specialRelease that will take in XK keys like XK_Return.
But see other fix below.

**Comment by lostallmymoney** [comment, marked as answer]
Try this :
1 - keyClick=XF86Copy
The keys are :
XF86Copy
XF86Paste
XF86Cut
It seems some letters in KeyList.txt file are not uppercase as they should.

We were messing around ahah !
We should not mix on press events and on release events, to make the deamon fully versatile I will need to add key descriptors to be able to make on press and on release run with synchronicity as right now the on release might run before the on press ends.

A user might want one behavior and another user might want the other behavior so possibly have both but that's a messy feature, better to plan ahead the keys on the press, or on the release for right now.

**Reply by giatmanos** [reply]
Shouldn't the last command be keyReleaseOnRelease?

**Reply by giatmanos** [reply]
This seems to work 100% of the few times I tested :)

**Reply by giatmanos** [reply]
But from what I understand, I would still expect a keyReleaseOnRelease to close the macro.

**Reply by giatmanos** [reply]
I tested with KeyReleaseOnRelease and it fails sometimes. keyReleaseOnPress at the end seems to be the solution.
Thanks!

**Reply by lostallmymoney** [reply]
No need to put anything on release ;) 

And yes I do understand the expected but if you were to have a thousand actions inside the press, and you would still want actions to run on release, then you wouldn't have to wait for them to end before running the on release.
This is in the optics you might be running commands by example. Might add a forced synchronicity tag on the keys one day so people can choose.

**Reply by lostallmymoney** [reply]
Hey anyways using xdotool I found the key it's 
	XF86Copy
	XF86Paste
	XF86Cut
It seems some letters in KeyList.txt file are not uppercase as they should.
1 -  keyClick=XF86Copy

**Reply by lostallmymoney** [reply]
Adding to the issue is the fact that by example the buttons on my mouse are unreliable and will sometime be hard to hit or in this case, will toggle while being help pressed. Which can mean repeating the same actions. Fixing this would mean waiting for an estimated and configurable time for the failures before doing any release action.
